### PR TITLE
DF-594: Correctly show 'no results' messaging when refined searches yield no results

### PR DESCRIPTION
**Pull request by ababic**
Ticket URL: https://national-archives.atlassian.net/browse/DF-594

## About these changes

As discussed on Slack, the 'buckets' were previously erroneously tied to the refined/filtered search results. The changes in #422 fixed this issue (tying them instead to the 'original' search query result), but failed to update the template logic conditions to render correctly under the right conditions.

## How to check these changes

If your original search query has no hits in any bucket, the buckets or refinement options should not display, and you should see a clear "No results" message.

Catalogue search example: http://0.0.0.0:8000/search/catalogue/?q=querty
Website search example: http://0.0.0.0:8000/search/website/?q=querty 

If your original search query has hits in at least one bucket, the buckets should display. However, if the current bucket has no hits, you should see a clear "No results" message and no further refinement options.

Catalogue search example: http://0.0.0.0:8000/search/catalogue/?q=snub&group=creator
Website search example: http://0.0.0.0:8000/search/website/?q=japan&group=highlight

If your original search query has hits in at least one bucket AND the 'current' bucket has results, the buckets should display. However, if you use the "Search within results" option and there are no results for that 'more refined' search, you should see a clear "No results" message. with no further refinement options (but you SHOULD be able to change/unapply the "Search within results" query

Catalogue search example: http://0.0.0.0:8000/search/catalogue/?q=japan&filter_keyword=querty
Website search example: http://0.0.0.0:8000/search/website/?q=japan&filter_keyword=querty

## Before assigning to reviewer, please make sure you have

- [x] Checked things thoroughly before handing over to reviewer.
- [x] Checked PR title starts with ticket number as per project conventions to help us keep track of changes.
- [x] Ensured that PR includes only commits relevant to the ticket.
- [x] Waited for all CI jobs to pass before requesting a review.
- [x] Added/updated tests and documentation where relevant.

## Merging PR guidance

Follow [docs\developer-guide\contributing.md](https://nationalarchives.github.io/ds-wagtail/developer-guide/contributing/)

## Deployment guidance

Follow [docs\infra\environments.md](https://nationalarchives.github.io/ds-wagtail/infra/environments/)

### CODE REVIEW

**Review by TNA-Allan — COMMENTED** [review]
Hi @ababic , when I ran the search query, the text "Try removing any filters that you may have applied" is missing. It appears templates\search\blocks\no_results_catalogue_website.html is used.

File: etna/search/templatetags/search_tags.py [hunk 1]
```diff
@@ -84,17 +84,6 @@ def include_hidden_fields(visible_field_names, form) -> str:
...
 
-@register.simple_tag(takes_context=True)
-def has_result_count(context, bucket) -> bool:
-    """Returns True if at least one bucket contains count returned by api response, otherwise False"""
-    buckets = context["buckets"]
-    for bucket in buckets:
-        if bucket.result_count:  # checks api has returned a count
-            if bucket.result_count > 0:
-                return True
-    return False
```
Comment (TNA-Allan) [inline etna/search/templatetags/search_tags.py:95]: The functionality to check  least one bucket is removed. What is that replaced with?

Comment (ababic) [reply etna/search/templatetags/search_tags.py:95]: It is no longer needed because we add `buckets_contain_results` to the template context

**Comment by ababic** [comment]
Hi @TNA-Allan. Regarding your point:

> When I ran the search query, the text "Try removing any filters that you may have applied" is missing. It appears templates\search\blocks\no_results_catalogue_website.html is used.

A better way to explain this might have been:

> On main, when using the catalogue search view and refining the search so that there are no results, the text "Try removing any filters that you may have applied" is included in the no results message. That appears to be missing when on this branch. This copy appears to be present in `templates\search\blocks\no_results_catalogue_website.html`

Anyhow, I figured out that the `search_results.html` include template was already handling "no results" cases, so I am letting that take care of the "No results in a refined search" case.

File: etna/search/tests/test_views.py [hunk 2]
```diff
@@ -223,6 +223,361 @@ def test_accessing_page_with_no_params_performs_empty_search(self):
...
+        )
+        content = str(response.content)
+
+        # SHOULD see
+        self.assertBucketLinksRendered(content)
+        self.assertSearchWithinOptionRendered(content)
+        self.assertSortOrderOptionsRendered(content)
+        self.assertFilterOptionsRendered(content)
+        self.assertNoResultsMessagingRendered(content)
+
```
Comment (TNA-Allan) [inline etna/search/tests/test_views.py:538]: The text in `no_results_messaging_html` ..._We did not find any results for your search_ is common across templates\search\blocks\no_results.html and templates\search\blocks\no_results_catalogue_website.html

Could we then have one which is specific for a Refined search with text "_Try removing any filters that you may have applied_" from `no_results_catalogue_website.html`

File: etna/search/tests/test_views.py [hunk 3]
```diff
@@ -223,6 +223,361 @@ def test_accessing_page_with_no_params_performs_empty_search(self):
...
+    def assertResultsNotRendered(self, response):
+        self.assertNotIn(self.results_html, response)
+
+
+class CatalogueSearchEndToEndTest(EndToEndSearchTestCase):
+    test_url = reverse_lazy("search-catalogue")
+
+    @responses.activate
+    def test_no_matches_for_original_search(self):
+        """
```
Comment (TNA-Allan) [inline etna/search/tests/test_views.py:301]:
```suggestion
    def test_no_matches_for_search_query_param_q(self):
        """
```
Not sure about specific terminology, but `_original_search` or `original search` seems weird. 
Passed on similar comment when using {# The original/unfiltered search failed to yield results #} 
Similar occurrences elsewhere `_original_search` or `original search`

File: templates/search/catalogue_search.html [hunk 4]
```diff
@@ -23,36 +23,33 @@
...
+        {% include './blocks/search_results_hero.html' %}
+    {% endwith  %}
 
-        {%  with search_tab=searchtabs.CATALOGUE.value%}
-            {% include './blocks/search_results_hero.html' %}
-        {% endwith  %}
-
-        {% if buckets.current.result_count > 0 %}
+    {% if not buckets_contain_results %}
+        {# The original/unfiltered search failed to yield results #}
```
[the remark this answers is not in the document]
Comment (ababic) [reply templates/search/catalogue_search.html]: For clarity, I have updated to read:
> No results were found when searching only with the 'q' param value

File: templates/search/catalogue_search.html [hunk 5]
```diff
@@ -23,36 +23,33 @@
...
-        {% comment %} TODO make this text an editable field {% endcomment %}
+        {% include './blocks/catalogue_search_buckets.html' %}
 
-        {% if buckets.current.description and buckets.current.result_count > 0 %}
-            <p class="search-results__explainer">
-                {{ buckets.current.description }}
-            </p>
-        {% endif %}
+        {% if not buckets.current.result_count %}
+            {% include './blocks/no_results.html' %}
```
[the remark this answers is not in the document]
Comment (ababic) [reply templates/search/catalogue_search.html:43]: @TNA-Allan Please see this test here: https://github.com/nationalarchives/ds-wagtail/pull/729/files#diff-ba302027a99f729f34bfd34e5b35d62a15b0c03e04031e78803688cb042876f6R332-R364

File: etna/search/tests/test_views.py (hunk 2 above)
Comment (ababic) [reply etna/search/tests/test_views.py:538]: I'm trying to deliberately steer away from testing actual messaging, as that is not really a testing concern, but an editorial one.

File: etna/search/tests/test_views.py (hunk 3 above)
Comment (ababic) [reply etna/search/tests/test_views.py]: Adjusted in [1073ff6](https://github.com/nationalarchives/ds-wagtail/pull/729/commits/1073ff6f0f9ea500b39e566b08a924313a88f839)

File: etna/search/tests/test_views.py [hunk 6]
```diff
@@ -223,6 +223,361 @@ def test_accessing_page_with_no_params_performs_empty_search(self):
         )
 
 
+@override_settings(KONG_CLIENT_BASE_URL="https://kong.test")
+class EndToEndSearchTestCase(TestCase):
```
[the remark this answers is not in the document]
Comment (ababic) [reply etna/search/tests/test_views.py:227]: If other classes want to subclass this in future, they can. I don't see any reason for any other classes to use it unless they're making use of the additions it brings.

File: etna/search/tests/test_views.py (hunk 2 above)
Comment (TNA-Allan) [reply etna/search/tests/test_views.py:538]: This is great. Thanks @ababic

**Pull request merged**